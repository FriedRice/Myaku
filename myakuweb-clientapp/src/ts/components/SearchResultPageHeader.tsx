/** @module Header component for a page of search results */

import React from 'react';
import Tile from './Tile';

interface SearchResultPageHeaderProps {
    totalResults: number;
    pageNum: number | null;
}
type Props = SearchResultPageHeaderProps;


const SearchResultPageHeader: React.FC<Props> = function(props) {
    var pageNumElement: React.ReactElement | null = null;
    if (props.pageNum !== null) {
        pageNumElement = (
            <small className='result-header-page-number'>
                {`Page ${props.pageNum.toLocaleString('en-US')}`}
            </small>
        );
    }

    return (
        <Tile tileClasses='results-header-tile'>
            <h3>
                <span>
                    {'Articles '}
                    <small>
                        {'â€” '}
                        {props.totalResults.toLocaleString('en-US')}
                        {' found'}
                    </small>
                </span>
                {pageNumElement}
            </h3>
        </Tile>
    );
};

export default SearchResultPageHeader;
