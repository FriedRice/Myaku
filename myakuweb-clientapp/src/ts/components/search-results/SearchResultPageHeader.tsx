/**
 * SearchResultPageHeader component module. See [[SearchResultPageHeader]].
 */

import React from 'react';
import Tile from 'ts/components/generic/Tile';

/** Props for the [[SearchResultPageHeader]] component. */
interface SearchResultPageHeaderProps {
    /** Total result count for the search */
    totalResults: number | null;

    /** Page number of the search results currently displaying */
    pageNum: number | null;
}
type Props = SearchResultPageHeaderProps;


/**
 * Get the total results element portion of the header.
 *
 * @param totalResults - Total result count for the search.
 *
 * @returns If the given total results is non-null, returns a text element
 * containing the given total results number.
 * If the given total results is null, returns a loading inline-block element.
 */
function getTotalResultsElement(
    totalResults: number | null
): React.ReactNode {
    if (totalResults === null) {
        var style: React.CSSProperties = {
            width: '5em',
            height: '1em',
            display: 'inline-block',
        };
        return <div className='loading' style={style}></div>;
    }

    return `${totalResults.toLocaleString('en-US')} found`;
}

/**
 * Get the page number element portion of the header.
 *
 * @param pageNum - Page number of the search results currently displaying.
 * @param totalResults - Total result count for the search.
 *
 * @returns If pageNum is non-null and totalResult is greater than 0, returns a
 * text element containing the given page number.
 * If pageNum is null or totalResults is not greater than 0, returns null.
 */
function getPageNumElement(
    pageNum: number | null, totalResults: number | null
): React.ReactNode {
    if (pageNum === null || totalResults === 0) {
        return null;
    }

    var classList = ['result-header-page-number'];
    return (
        <small className={classList.join(' ')}>
            {`Page ${pageNum.toLocaleString('en-US')}`}
        </small>
    );
}

/**
 * Header component for a page of search results
 *
 * @param props - See [[SearchResultPageHeaderProps]].
 */
const SearchResultPageHeader: React.FC<Props> = function(props) {
    return (
        <Tile tileClasses='results-header-tile'>
            <h3>
                <span>
                    {'Articles '}
                    <small>
                        {'â€” '}
                        {getTotalResultsElement(props.totalResults)}
                    </small>
                </span>
                {getPageNumElement(props.pageNum, props.totalResults)}
            </h3>
        </Tile>
    );
};

export default SearchResultPageHeader;
