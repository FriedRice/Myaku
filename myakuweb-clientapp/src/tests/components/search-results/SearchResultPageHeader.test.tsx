/**
 * Tests for the [[SearchResultPageHeader]] component.
 */

import React from 'react';
import SearchResultPageHeader from
    'ts/components/search-results/SearchResultPageHeader';

import {
    render,
    shallow,
} from 'enzyme';


describe('<SearchResultPageHeader />', function() {
    it(
        'renders correctly with null total results/null page num',
        function() {
            const wrapper = shallow(
                <SearchResultPageHeader
                    totalResults={null}
                    pageNum={null}
                />
            );
            expect(wrapper).toMatchSnapshot();
        }
    );

    it(
        'renders correctly with non-null total results/null page num',
        function() {
            const wrapper = shallow(
                <SearchResultPageHeader
                    totalResults={155}
                    pageNum={null}
                />
            );
            expect(wrapper).toMatchSnapshot();
        }
    );

    it(
        'renders correctly with null total results/non-null page num',
        function() {
            const wrapper = shallow(
                <SearchResultPageHeader
                    totalResults={null}
                    pageNum={1}
                />
            );
            expect(wrapper).toMatchSnapshot();
        }
    );

    it(
        'renders correctly with non-null total results/non-null page num',
        function() {
            const wrapper = shallow(
                <SearchResultPageHeader
                    totalResults={155}
                    pageNum={1}
                />
            );
            expect(wrapper).toMatchSnapshot();
        }
    );

    it('does not show page num if 0 total results', function() {
        const wrapper = shallow(
            <SearchResultPageHeader
                totalResults={0}
                pageNum={1}
            />
        );
        expect(wrapper).toMatchSnapshot();
    });

    it('uses int comma for total results and page num', function() {
        const wrapper = render(
            <SearchResultPageHeader
                totalResults={3444555}
                pageNum={1222}
            />
        );
        expect(wrapper.text()).toStrictEqual(
            expect.stringContaining('3,444,555 found')
        );
        expect(wrapper.text()).toStrictEqual(
            expect.stringContaining('Page 1,222')
        );
    });
});
