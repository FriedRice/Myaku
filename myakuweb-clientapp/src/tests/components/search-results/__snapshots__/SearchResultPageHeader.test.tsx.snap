// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchResultPageHeader /> does not show page num if 0 total results 1`] = `
<Tile
  tileClasses="results-header-tile"
>
  <h3>
    <span>
      Articles 
      <small>
        — 
        0 found
      </small>
    </span>
  </h3>
</Tile>
`;

exports[`<SearchResultPageHeader /> renders correctly with non-null total results/non-null page num 1`] = `
<Tile
  tileClasses="results-header-tile"
>
  <h3>
    <span>
      Articles 
      <small>
        — 
        155 found
      </small>
    </span>
    <small
      className="result-header-page-number"
    >
      Page 1
    </small>
  </h3>
</Tile>
`;

exports[`<SearchResultPageHeader /> renders correctly with non-null total results/null page num 1`] = `
<Tile
  tileClasses="results-header-tile"
>
  <h3>
    <span>
      Articles 
      <small>
        — 
        155 found
      </small>
    </span>
  </h3>
</Tile>
`;

exports[`<SearchResultPageHeader /> renders correctly with null total results/non-null page num 1`] = `
<Tile
  tileClasses="results-header-tile"
>
  <h3>
    <span>
      Articles 
      <small>
        — 
        <div
          className="loading"
          style={
            Object {
              "display": "inline-block",
              "height": "1em",
              "width": "5em",
            }
          }
        />
      </small>
    </span>
    <small
      className="result-header-page-number"
    >
      Page 1
    </small>
  </h3>
</Tile>
`;

exports[`<SearchResultPageHeader /> renders correctly with null total results/null page num 1`] = `
<Tile
  tileClasses="results-header-tile"
>
  <h3>
    <span>
      Articles 
      <small>
        — 
        <div
          className="loading"
          style={
            Object {
              "display": "inline-block",
              "height": "1em",
              "width": "5em",
            }
          }
        />
      </small>
    </span>
  </h3>
</Tile>
`;
